Exclude muted users from more UI parts.

abhijeetbodas2001 opened the pull request:
This deals with-
- Typing notifications
- PM recipients typeahead
- @ mentions typeahead

Tested manually, and also added node tests.

abhijeetbodas2001 commented:
@m-e-l-u-h-a-n Since you worked on compose typeahead recently, could you review this?

m-e-l-u-h-a-n commented on static/js/user_pill.js:
```diff
@@ -94,7 +95,8 @@ export function has_unconverted_data(pill_widget) {
 }
 
 export function typeahead_source(pill_widget) {
-    const persons = people.get_realm_users();
+    let persons = people.get_realm_users();
+    persons = persons.filter((person) => !muting.is_user_muted(person.user_id));
```
This will only remove it, recipient typeahead but pills for muted users can still be created by pasting email of users in recipient box. e.g.: `user7@zulipdev.com, user6@zulipdev.com`.
If we want to prevent this then we can modify `create_item_from_email` function in `user_pill.js` to return undefined if the user is muted but this will have some other implications wherever user_pill.js is used for creating user pills.

m-e-l-u-h-a-n replied:
If User A has muted User B then this typeahead source change and above suggested copy-paste change will affect other places too:
* User A cannot add User B to any user group(through UI).
* User A cannot add User B to stream using stream-edit UI.
* User A cannot create pill for User B in recipient box(which means cannot send private message).

abhijeetbodas2001 replied:
I don't think we want to totally *prevent* A from adding B in any input, but to just make sure B does not clutter up the typeahead results for A. Maybe I should be writing this in the commit message?

timabbott replied:
Let's make a `people.get_active_unmuted_users` function and call that from all of these code paths -- that's going to avoid duplicating this filter call in N places.

timabbott replied:
On the functionality question, @WesleyAC @alya thoughts on which of "Add subscribers to stream", "Compose box typeahead", and "User groups typeahead" should allow muted users?  I think the policy we'd discussed had been that only "Compose box typeahead" should hide them.

abhijeetbodas2001 replied:
The [doc](https://docs.google.com/document/d/1UFQnai5t8R6Fp8lDiWlO9XWuqTUrMrI1tMd-7e7ml3Y/edit#heading=h.vro3nqktoc7j) just says "typeahead lists". Maybe we should update that too?

timabbott replied:
Yeah, I meant "compose typeahead"; we should update that in the doc.

alya replied:
Yeah, I think it's nice to have a consistent pattern of showing muted users in the menus, including "Add subscribers to stream" and "User groups typeahead".

Removing muted users from "Compose box typeahead" makes sense to me as well.

m-e-l-u-h-a-n commented:
I went through the changes, and posted a couple of minor comment changes to `user_pill.js` rest of the changes appear fine 🙂️.

abhijeetbodas2001 commented:
Thanks a lot for the review! I replied with the reasoning inline, and also mentioned it in the commit message.
I also added a commit here for the stream edit typeahead, which I had missed earlier.

timabbott commented on static/js/typing_events.js:
```diff
@@ -52,7 +53,9 @@ function get_users_typing_for_narrow() {
 export function render_notifications_for_narrow() {
-    const user_ids = get_users_typing_for_narrow();
+    let user_ids = get_users_typing_for_narrow();
+    user_ids = user_ids.filter((id) => !muting.is_user_muted(id));
```
I think it'd be better to push this down into the `typing_data.js` logic -- we can rename `get_all_typists` to `get_unmuted_typists` and do the check there.

abhijeetbodas2001 replied:
Done!

timabbott commented on static/js/stream_edit.js:
```diff
@@ -387,8 +388,10 @@ function show_subscription_settings(sub) {
+        let potential_subscribers = peer_data.potential_subscribers(stream_id);
+        potential_subscribers = user_pill.filter_taken_users(potential_subscribers, pill_widget);
+        // Exclude muted users from add-subscriber typeahead.
+        return potential_subscribers.filter((user) => !muting.is_user_muted(user.user_id));
```
I'm not sure we want to exclude the muted user from this widget, since it's part of the settings interface?

abhijeetbodas2001 replied:
Ok, dropped this commit.

abhijeetbodas2001 commented:
Thanks for reviewing! Changes in the last push are-
- Removed the couple of commits dealing with the stream/user group settings pages.
- Added some helpers in `muting.js` for filtering muted users so that we don't duplicate the same code (and also added a commit to clean up code from buddy data and pm_conversations which had already been merged). I preferred this approach over something like `people.get_active_unmuted_users()` because many modules use custom data sources, rather than directly doing `people.get_active_users()`.
- The typing logic is now in `typing_data` instead of `typing_events` and also covered by node tests 🙂 
- The first two commits are kind of unrelated to the rest of the PR now, but I didn't drop them because I think they are a good cleanup.

timabbott commented:
This looks good, except maybe we should rename the functions to `filter_users_for_muting` and `filter_user_ids_for_muting`?  `get` I usually expect to not return a subset of what was passed in, but instead query somewhere else.

abhijeetbodas2001 commented:
Changed the function names to `filter_muted_users` and `filter_muted_user_ids`

timabbott commented on static/js/buddy_data.js:
```diff
@@ -327,14 +327,12 @@ function filter_user_ids(user_filter_text, user_ids) {
     });
 
+    // Muted users are hidden from the right sidebar entirely.
+    user_ids = muting.filter_muted_user_ids(user_ids);
```
I think we don't want to do this particular refactor, basically because `user_ids` is a huge list that we rerender frequently, and it is probably worth avoiding unnecessary copies of the list.

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -449,6 +450,10 @@ export function get_person_suggestions(query, opts) {
         }
+
+        // Exclude muted users from typeaheads.
+        persons = muting.filter_muted_users(persons);
```
I think we probably want this to be before `broadcast_mentions`, since those aren't people.

timabbott commented:
Merged as the series ending with 65249d90e9fc2cac8fbf8b745a5a7012eef90180, thanks @abhijeetbodas2001!

I think we're almost there!
